contracts/button/lib.rs

```diff
@@ -1,390 +1,377 @@
+            let required_role = Role::Owner(this);
+            self.check_role(caller, required_role)?;
+            self.env().terminate_contract(caller)
         }
 
-        // reward the Pressiah
-        if let Some(pressiah) = last_presser {
-            let reward = self.pressiah_score();
-            self.mint_tx::<E>(*pressiah, reward)??;
-        };
+        //===================================================================================================
+
+        fn init(
+            ticket_token: AccountId,
+            reward_token: AccountId,
+            button_lifetime: BlockNumber,
+            scoring: Scoring,
+        ) -> Self {
+            let now = Self::env().block_number();
+            let deadline = now + button_lifetime;
+
+            let contract = Self {
+                access_control: AccountId::from(ACCESS_CONTROL_PUBKEY),
+                button_lifetime,
+                reward_token,
+                ticket_token,
+                last_press: now,
+                scoring,
+                last_presser: None,
+                presses: 0,
+                total_rewards: 0,
+            };
+
+            Self::emit_event(
+                Self::env(),
+                Event::ButtonCreated(ButtonCreated {
+                    start: now,
+                    deadline,
+                    ticket_token,
+                    reward_token,
+                }),
+            );
+
+            contract
+        }
 
-        // zero the counters in storage
-        let mut state = self.get_mut();
+        fn check_role(&self, account: AccountId, role: Role) -> ButtonResult<()>
+        where
+            Self: AccessControlled,
+        {
+            <Self as AccessControlled>::check_role(
+                self.access_control,
+                account,
+                role,
+                |why: InkEnvError| {
+                    GameError::ContractCall(format!("Calling access control has failed: {:?}", why))
+                },
+                |role: Role| GameError::MissingRole(format!("{:?}", role)),
+            )
+        }
 
-        state.presses = 0;
-        state.last_presser = None;
-        state.last_press = now;
-        state.total_rewards = 0;
+        fn score(&self, now: BlockNumber) -> Balance {
+            match self.scoring {
+                Scoring::EarlyBirdSpecial => self.deadline().saturating_sub(now) as Balance,
+                Scoring::BackToTheFuture => now.saturating_sub(self.last_press) as Balance,
+                Scoring::ThePressiahCometh => (self.presses + 1) as Balance,
+            }
+        }
 
-        Ok(())
-    }
-}
+        fn pressiah_score(&self) -> Balance {
+            (self.total_rewards / 4) as Balance
+        }
+
+        fn transfer_ticket(
+            &self,
+            from: AccountId,
+            to: AccountId,
+            value: Balance,
+        ) -> Result<Result<(), PSP22Error>, InkEnvError> {
+            build_call::<DefaultEnvironment>()
```

Comment:
we should make a lib of all these cross contract calls

Reply:
https://cardinal-cryptography.atlassian.net/browse/A0-1380